### PR TITLE
[mysql] Let the records of snapshot split don't cross checkpoints

### PR DESCRIPTION
**Pull request by leonardBang**
The output records of a snapshot split should not cross checkpoints, this may lead to data duplication when failover happen, because partial snapshot records of one split may sent in last checkpoint, but this snapshot split may be re-read all snapshot records after failover, this behavior bring potential data duplication issue.

### CODE REVIEW

**Comment by leonardBang** [comment]
@ruanhang1993 Would you like to help review this PR？

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceITCase.java [hunk 1]
```diff
@@ -264,6 +292,132 @@ public void testConsumingTableWithoutPrimaryKey() {
...
+        // Trigger failover once some snapshot records has been sent by sleeping source
+        if (iterator.hasNext()) {
+            triggerFailover(
+                    FailoverType.JM,
+                    jobId,
+                    miniClusterResource.getMiniCluster(),
+                    () -> sleepMs(100));
+        }
+
+        List<String> expectedSnapshotData = new ArrayList<>(Arrays.asList(snapshotForSingleTable));
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceITCase.java:361]: Rename `snapshotForSingleTable` to `expectedSnapshotData` and remove this line.

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReaderTest.java [hunk 2]
```diff
@@ -381,4 +467,127 @@ public TypeInformation<SourceRecord> getProducedType() {
             return TypeInformation.of(SourceRecord.class);
         }
     }
+
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReaderTest.java:470]: Add some descriptions about this emitter.

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReaderTest.java [hunk 3]
```diff
@@ -136,7 +156,67 @@ public void testBinlogReadFailoverCrossTransaction() throws Exception {
...
                 };
         // the 4 records are produced by 3 operations
-        List<String> restRecords = consumeRecords(restartReader, dataType, 3);
+        List<String> restRecords = consumeRecords(restartReader, dataType);
+        assertEqualsInOrder(Arrays.asList(expectedRestRecords), restRecords);
+        restartReader.close();
+    }
+
+    @Test
+    public void testBinlogReadFailoverCrossCheckpoint() throws Exception {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReaderTest.java:165]: The code in `testBinlogReadFailoverCrossTransaction` and `testBinlogReadFailoverCrossCheckpoint` is the same.

**Review by ruanhang1993 — COMMENTED** [review]
@leonardBang Thanks for the contribution. I left some comments.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java [hunk 4]
```diff
@@ -155,7 +156,9 @@ public Iterator<SourceRecord> pollSplitRecords() throws InterruptedException {
                     sourceRecords.add(event.getRecord());
                 }
             }
-            return sourceRecords.iterator();
+            Set<SourceRecords> sourceRecordsSet = new HashSet<>();
+            sourceRecordsSet.add(new SourceRecords(sourceRecords));
+            return sourceRecordsSet.iterator();
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java:161]: this is the legacy code, I'll fix it

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceITCase.java (hunk 1 above)
Comment (leonardBang) [reply flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceITCase.java:361]: assertEqualsInAnyOrder accept `List`, I'll improve this piece

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 5]
```diff
@@ -288,7 +291,10 @@ public Iterator<SourceRecord> pollSplitRecords() throws InterruptedException {
             normalizedRecords.add(lowWatermark);
             normalizedRecords.addAll(formatMessageTimestamp(snapshotRecords.values()));
             normalizedRecords.add(highWatermark);
-            return normalizedRecords.iterator();
+
+            final Set<SourceRecords> sourceRecordsSet = new HashSet<>();
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:295]: I see that we use `List` in `BinlogSplitReader`. So we better use `List` here as well.

**Review by ruanhang1993 — APPROVED** [review]
LGTM. Thanks. @leonardBang

**Pull request merged**